File: V2/files (2)/domain-pipeline-schema/src/nodes/node2.ts
/**
 * Node 2 — Aggregate & Statechart Extraction
 *
 * Clusters events into aggregates, extracts implicit state machines,
 * performs gap analysis. State discovery carries evidence (partial ReAct);
 * statechart construction is analytical reasoning over validated Node 1 data.
 *
 * Validation layers:
 *   L1 (schema): field presence, ID formats, enum values
 *   L2 (referential): aggregate codes match registry; transition.event
 *                      resolves to Node 1 catalogue; gap ID prefixes match aggregate
 *   L3 (semantic): no dead-end states without justification;
 *                   every event referenced; gap severity distribution sensible;
 *                   aggregate codes unique; state names unique within aggregate
 */

import { z } from "zod";
import {
  PipelineEnvelope,
  AggCode,
  EventID,
  GapID,
  CodeRef,
  Severity,
  StateType,
  TransitionAnnotation,
} from "../primitives.js";

// ── State ───────────────────────────────────────────────────────────────────

export const StateEntry = z.object({
  name: z.string().min(1),
  type: StateType,
  representation: z
    .string()
    .min(1, "How this state appears in code (enum value, boolean combo, etc)"),
  location: CodeRef,

  /** ReAct grounding for state discovery: what code pattern reveals this state. */
  evidence: z.string().min(10, "State must be evidenced from code"),
});
export type StateEntry = z.infer<typeof StateEntry>;

// ── Transition ──────────────────────────────────────────────────────────────

export const Transition = z.object({
  source: z.string().min(1),
  target: z.string().min(1),
  event: EventID,
  guard: z.string().nullable(),
  sideEffects: z.array(z.string()),
  annotation: TransitionAnnotation.nullable(),
});
export type Transition = z.infer<typeof Transition>;

// ── Gap ─────────────────────────────────────────────────────────────────────

export const Gap = z.object({
  id: GapID,
  severity: Severity,
  description: z.string().min(1),
  codeLocation: CodeRef,
});
export type Gap = z.infer<typeof Gap>;

// ── Aggregate ───────────────────────────────────────────────────────────────

export const Aggregate = z.object({
  code: AggCode,
  name: z.string().min(1),
  rootEntity: z.string().min(1),
  keyFiles: z.array(CodeRef).min(1),
  states: z.array(StateEntry).min(1),
  transitions: z.array(Transition),
  gaps: z.array(Gap),
  /** If ≤2 states: "trivial lifecycle". Full analysis only if 3+. */
  trivialLifecycle: z.boolean().default(false),
});
export type Aggregate = z.infer<typeof Aggregate>;

// ── Cross-aggregate structures ──────────────────────────────────────────────

export const CrossAggTransition = z.object({
  fromEvent: EventID,
  fromAggregate: AggCode,
  toEvent: EventID,
  toAggregate: AggCode,
  mechanism: z.string().min(1),
});

export const ImpossibleCombination = z.object({
  combination: z.string().min(1),
  risk: z.string().min(1),
  source: z.string().min(1),
});

export const MigrationCandidate = z.object({
  rank: z.number().int().min(1),
  aggregate: AggCode,
  rationale: z.string().min(1),
});

// ── Node 2 Output ───────────────────────────────────────────────────────────

export const Node2Output = PipelineEnvelope.extend({
  node: z.literal(2),
  aggregates: z.array(Aggregate).min(1),
  crossAggTransitions: z.array(CrossAggTransition),
  impossibleCombinations: z.array(ImpossibleCombination),
  migrationPriority: z.array(MigrationCandidate),
  gapSummary: z.object({
    critical: z.number().int().min(0),
    high: z.number().int().min(0),
    med: z.number().int().min(0),
    low: z.number().int().min(0),
    total: z.number().int().min(0),
  }),
});

export type Node2Output = z.infer<typeof Node2Output>;

File: v3_python/ede/nodes/node2.py
"""
Node 2 -- Aggregate & Statechart Extraction

Clusters events into aggregates, extracts implicit state machines,
performs gap analysis.

Validation layers:
  L1 (schema): field presence, ID formats, enum values
  L2 (referential): aggregate codes match registry; transition.event
                     resolves to Node 1 catalogue; gap ID prefixes match aggregate
  L3 (semantic): no dead-end states without justification;
                  gap severity distribution sensible
"""

from __future__ import annotations

from typing import Literal, Self

from pydantic import Field, model_validator

from ..primitives import (
    AggCode,
    CodeRef,
    EdeBaseModel,
    EventID,
    GapID,
    PipelineEnvelope,
    Severity,
    StateType,
    TransitionAnnotation,
)


class StateEntry(EdeBaseModel):
    name: str = Field(min_length=1)
    type: StateType
    representation: str = Field(min_length=1)
    location: CodeRef
    evidence: str = Field(min_length=10)


class Transition(EdeBaseModel):
    source: str = Field(min_length=1)
    target: str = Field(min_length=1)
    event: EventID
    guard: str | None
    side_effects: list[str]
    annotation: TransitionAnnotation | None
    evidence: str = Field(min_length=10)
    code_location: CodeRef


class Gap(EdeBaseModel):
    id: GapID
    severity: Severity
    description: str = Field(min_length=1)
    code_location: CodeRef


class Aggregate(EdeBaseModel):
    code: AggCode
    name: str = Field(min_length=1)
    root_entity: str = Field(min_length=1)
    key_files: list[CodeRef] = Field(min_length=1)
    states: list[StateEntry] = Field(min_length=1)
    transitions: list[Transition]
    gaps: list[Gap]
    trivial_lifecycle: bool = False


class CrossAggTransition(EdeBaseModel):
    from_event: EventID
    from_aggregate: AggCode
    to_event: EventID
    to_aggregate: AggCode
    mechanism: str = Field(min_length=1)


class ImpossibleCombination(EdeBaseModel):
    combination: str = Field(min_length=1)
    risk: str = Field(min_length=1)
    source: str = Field(min_length=1)


class MigrationCandidate(EdeBaseModel):
    rank: int = Field(ge=1)
    aggregate: AggCode
    rationale: str = Field(min_length=1)


class GapSummary(EdeBaseModel):
    """L1 promotion: total must equal sum of components."""

    critical: int = Field(ge=0)
    high: int = Field(ge=0)
    med: int = Field(ge=0)
    low: int = Field(ge=0)
    total: int = Field(ge=0)

    @model_validator(mode="after")
    def check_total(self) -> Self:
        expected = self.critical + self.high + self.med + self.low
        if self.total != expected:
            raise ValueError(f"total ({self.total}) != sum of components ({expected})")
        return self


class Node2Output(PipelineEnvelope):
    node: Literal[2]
    aggregates: list[Aggregate] = Field(min_length=1)
    cross_agg_transitions: list[CrossAggTransition]
    impossible_combinations: list[ImpossibleCombination]
    migration_priority: list[MigrationCandidate]
    gap_summary: GapSummary
